test/rspamd_cxx_unit_cryptobox.hxx

```diff
@@ -0,0 +1,182 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+/* Detached unit tests for the cryptobox */
+
+#ifndef RSPAMD_RSPAMD_CXX_UNIT_CRYPTOBOX_HXX
+#define RSPAMD_RSPAMD_CXX_UNIT_CRYPTOBOX_HXX
+#include "libcryptobox/cryptobox.h"
+
+TEST_SUITE("rspamd_cryptobox")
+{
+	/*
+	TEST_CASE("rspamd_cryptobox_keypair")
+	{
+		enum rspamd_cryptobox_mode mode = RSPAMD_CRYPTOBOX_MODE_NIST;
+		rspamd_sk_t sk;
+		rspamd_pk_t pk;
+
+
+		ottery_rand_bytes(sk, sizeof(sk));
+		ottery_rand_bytes(pk, sizeof(pk));
+
+		rspamd_cryptobox_keypair(pk, sk, mode);
+	}
+
+	TEST_CASE("rspamd_cryptobox_keypair_sig")
+	{
+		enum rspamd_cryptobox_mode mode = RSPAMD_CRYPTOBOX_MODE_NIST;
+		rspamd_sig_sk_t sk;
+		rspamd_sig_pk_t pk;
+
+
+		ottery_rand_bytes(sk, sizeof(sk));
+		ottery_rand_bytes(pk, sizeof(pk));
+
+		rspamd_cryptobox_keypair_sig(pk, sk, mode);
+	}
+*/
+	TEST_CASE("rspamd_cryptobox_hash")
+	{
+		rspamd_cryptobox_hash_state_t p;
+		const unsigned char *key = reinterpret_cast<const unsigned char *>("keys");
+		gsize keylen = sizeof(key);
+
+		ottery_rand_bytes(&p, rspamd_cryptobox_HASHSTATEBYTES);
+
+		rspamd_cryptobox_hash_init(&p, key, keylen);
+
+		const unsigned char* data = reinterpret_cast<const unsigned char *>("data");
+		gsize len = sizeof(data);
+
+		rspamd_cryptobox_hash_update(&p, data, len);
+
+		unsigned char out1[rspamd_cryptobox_HASHSTATEBYTES];
+		memset(out1, 0, rspamd_cryptobox_HASHSTATEBYTES);
+
+		rspamd_cryptobox_hash_final(&p, out1);
+		CHECK(*out1 == 'd');
+
+		unsigned char out2[rspamd_cryptobox_HASHSTATEBYTES];
+		memset(out2, 0, rspamd_cryptobox_HASHSTATEBYTES);
+
+		rspamd_cryptobox_hash(out2, data, len, key, keylen);
+		CHECK(*out1 == *out2);
+	}
+
+	TEST_CASE("rspamd_cryptobox_fast_hash")
+	{
+		rspamd_cryptobox_fast_hash_state_s *st = rspamd_cryptobox_fast_hash_new();
+
+		uint64_t seed = 10;
+
+		rspamd_cryptobox_fast_hash_init(st, seed);
+
+		const unsigned char* data = reinterpret_cast<const unsigned char *>("data");
+		gsize len = sizeof(data);
+
+		rspamd_cryptobox_fast_hash_update(st, data, len);
+
+		uint64_t out1 = rspamd_cryptobox_fast_hash_final(st);
+
+		uint64_t out2 = rspamd_cryptobox_fast_hash(data, len, seed);
+		CHECK(out1 == out2);
+
+		rspamd_cryptobox_fast_hash_free(st);
+	}
+
+	TEST_CASE("rspamd_cryptobox_pbkdf")
+	{
+		const char *pass = "passpa";
+		gsize pass_len = sizeof(pass);
+
+		const uint8_t salt = 100;
+		gsize salt_len = sizeof(salt);
+
+		uint8_t key = 20;
+		gsize key_len = sizeof(key);
+
+		unsigned int complexity = 10;
+		enum rspamd_cryptobox_pbkdf_type type = RSPAMD_CRYPTOBOX_PBKDF2;
+
+
+		CHECK(rspamd_cryptobox_pbkdf(pass, pass_len, &salt, salt_len, &key, key_len, complexity, type) == true);
+		CHECK(key == 60);
+
+		type = RSPAMD_CRYPTOBOX_CATENA;
+		CHECK(rspamd_cryptobox_pbkdf(pass, pass_len, &salt, salt_len, &key, key_len, complexity, type) == true);
+		CHECK(key == 204);
```

Comment:
Again, this is not enough: we need to ensure that pbkdf works correctly for a given password/salt pair.

Reply:
pbkdf returns a string that is a safe to store representation of the password. So we need to check if we can derive the same encrypted string from the same password given that we use the same salt. Also, why salt is only 1 byte?